api: Add a way to play a sound from a coordinate

kezz opened the pull request:
This is an example implementation of #317 using a method similar to the `RGBLike` interface used to allow the use of native platform colour objects.

Closes #317.

kashike commented on api/src/main/java/net/kyori/adventure/util/CoordinateLike.java:
```diff
@@ -0,0 +1,68 @@
+ *
+ * @since 4.8.0
+ */
+public interface CoordinateLike {
```
```suggestion
public interface Vector3dLike {
```

kashike commented on api/src/main/java/net/kyori/adventure/util/CoordinateLikeImpl.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @since 4.8.0
+ */
+class CoordinateLikeImpl implements CoordinateLike {
```
```suggestion
final class Vector3dImpl implements Vector3dLike {
```

kashike replied:
Requires `toString` (using examination), `equals`, and `hashCode`.

kashike commented on api/src/main/java/net/kyori/adventure/audience/Audience.java:
```diff
@@ -438,6 +439,17 @@ default void hideBossBar(final @NonNull BossBar bar) {
   }
 
+  /**
+   * Plays a sound at a location.
+   *
+   * @param sound a sound
+   * @param location a location
+   * @since 4.8.0
+   */
+  default void playSound(final @NonNull Sound sound, final @NonNull Vector3dLike location) {
+    this.playSound(sound, location.x(), location.y(), location.z());
```
```suggestion
   * Plays a sound at a position.
   *
   * @param sound a sound
   * @param pos a position
   * @since 4.8.0
   */
  default void playSound(final @NonNull Sound sound, final @NonNull Vector3dLike pos) {
    this.playSound(sound, pos.x(), pos.y(), pos.z());
```

zml2008 commented on api/src/main/java/net/kyori/adventure/util/Vector3dLike.java:
```diff
@@ -0,0 +1,82 @@
+   * @return a new CoordinateLike
+   * @since 4.8.0
+   */
+  static Vector3dLike of(final double x, final double y, final double z) {
```
I don't know if adventure itself should provide an implementation of this interface.

kezz replied:
There's essentially zero chance that anyone implementing Adventure won't provide an object that can implement this. I suppose it can't hurt but happy to remove this.

MiniDigger commented:
if we add Vector3dLike, shouldn't we also provide Vector3iLike?

kezz commented:
> if we add Vector3dLike, shouldn't we also provide Vector3iLike?

It's still `x, y, z` right? We could generify this into just `Vector3Like`, that's how Unity names their corresponding object for example.

kashike commented:
> > if we add Vector3dLike, shouldn't we also provide Vector3iLike?
> 
> It's still `x, y, z` right? We could generify this into just `Vector3Like`, that's how Unity names their corresponding object for example.

No. Different return types.

kezz commented:
> Different return types.
Ah right. In that case yes I agree.

kezz commented:
I've just rebased this with the addition of `Vector3iLike` and the removal of the implementation for `Vector3dLike`. However, this does need further consideration as to how we would or if this PR even should handle worlds. Many platforms have objects that represent a location in a given world and this could produce unexpected behavior with code like this (using Paper as an example):

```java
Location location = new Location(Bukkit.getWorld("test"), 0.0, 0.0, 0.0);
Bukkit.getServer().playSound(sound, location); // why can't i do this? or why is it playing for people not in the "test" world?
```

I would be interested to hear everyone's thoughts on this, but I feel like the best way is what I have just added in this PR which is to add *another* location-ish object that does not provide any location in order to let platforms implement this as they feel fit. Would be happy to pull this last commit into another PR if necessary.

Additionally, the current implementation of the `playSound` method that takes a `Vector3iLike` just turns each integer into a double directly, should this be instead using the center of a location (i.e. adding 0.5 to each integer)?

zml2008 commented on api/src/main/java/net/kyori/adventure/util/LocationLike.java:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @since 4.8.0
+ */
+public interface LocationLike {
```
not a fan of an empty interface here -- and tbh not sure how important it is to have a worldly variant of `playSound`

kezz commented:
Been thinking about this one a bit more, and I'm just going to close this PR. I haven't seen any need for something like this in Adventure yet (nor has anyone else expressed an interest). Stateful API (that requires the developer to have information about the platform at hand, such as location, world, etc.) is perhaps just not as useful in Adventure anyway.
